ambry-api/src/main/java/com/github/ambry/frontend/UrlSigningService.java

```diff
@@ -38,9 +41,21 @@
   boolean isRequestSigned(RestRequest restRequest);
 
   /**
-   * Verifies that the signature in {@code restRequest} is valid.
+   * Verifies that the signature in {@code restRequest} is valid. Any remote calls in the implementation should be made
+   * asynchronously.
    * @param restRequest the {@link RestRequest} to check.
-   * @throws RestServiceException if there are problems verifying the URL.
+   * @param callback the {@link Callback} that will be called after signature verification.
    */
-  void verifySignedRequest(RestRequest restRequest) throws RestServiceException;
+  void verifySignedRequest(RestRequest restRequest, Callback<Void> callback);
+
+  /**
+   * Similar to {@link #verifySignedRequest(RestRequest, Callback)}, but returns a future.
+   * @param restRequest the {@link RestRequest} to check.
+   * @return a {@link Future} that will be completed when signature verification is complete.
+   */
+  default Future<Void> verifySignedRequest(RestRequest restRequest) {
```

Comment:
Agreed, changed it to be backwards compatible, since most implementations won't have the async need anyways. I got rid of the future based method since the method signature collides with `void verifySignedRequest(RestRequest restRequest)`. I think that's okay since there is no need for that method if the blocking method still exists.